src/Ldap.php

```diff
@@ -1130,6 +1130,100 @@ public function search(
         return $this->createCollection($iterator, $collectionClass);
     }
 
+    
+    
+        /**
+     * An LDAP search routine for finding information and returning paginated results
+     * https://stackoverflow.com/questions/16892693/zf2-ldap-pagination
+     *
+     * Options can be either passed as single parameters according to the
+     * method signature or as an array with one or more of the following keys
+     * - filter
+     * - baseDn
+     * - scope
+     * - attributes
+     * - sort
+     * - collectionClass
+     * - sizelimit
+     * - timelimit
+     *
+     * @param  string|Filter\AbstractFilter|array $filter
+     * @param  string|Dn|null                     $basedn
+     * @param  int                                $scope
+     * @param  array                              $attributes
+     * @param  string|null                        $sort
+     * @param  string|null                        $collectionClass
+     * @param  integer                            $timelimit
+     * @param  integer                            $pageSize
+     * @return Array
+     * @throws Exception\LdapException
+     */
+    public function multiPageSearch(
+        $filter, $basedn = null, $scope, array $attributes = array(), $sort = null,
+        $collectionClass = null, $timelimit = 0, $pageSize = 10000
+    )
+    {
+        if (is_array($filter)) {
+            $options = array_change_key_case($filter, CASE_LOWER);
+            foreach ($options as $key => $value) {
+                switch ($key) {
+                    case 'filter':
+                    case 'basedn':
+                    case 'scope':
+                    case 'sort':
+                        $$key = $value;
+                        break;
+                    case 'attributes':
+                        if (is_array($value)) {
+                            $attributes = $value;
+                        }
+                        break;
+                    case 'collectionclass':
+                        $collectionClass = $value;
+                        break;
+                    case 'sizelimit':
+                    case 'timelimit':
+                        $$key = (int) $value;
+                        break;
+                }
+            }
+        }
+
+        if ($basedn === null) {
+            $basedn = $this->getBaseDn();
+        } elseif ($basedn instanceof Dn) {
+            $basedn = $basedn->toString();
+        }
+
+        if ($filter instanceof Filter\AbstractFilter) {
+            $filter = $filter->toString();
+        }
+
+        $resource = $this->getResource();
+        ErrorHandler::start(E_WARNING);
+        $cookie = '';
+        $results = [];
+        do {
+            ldap_control_paged_result($resource, $pageSize, true, $cookie);
+
+             $result  = ldap_search($resource, $basedn,  $filter,
+                 $attributes
+             );
+
+            foreach (ldap_get_entries($resource, $result) as $item){
```

Comment:
I see an issue here as `ldap_get_entries` returns a rather awkward array containing an entry `count` followed by a numerically indext list of entries. So for each iteration in the `do...while` we will have an entry `count` pushed to the `results` array. I doubt that is what you had in mind...